File: runtime/Gaffer.Runtime/Events/FeedResult.cs
using Gaffer.Sdk.Diagnostics;

namespace Gaffer.Runtime.Events;

/// <summary>Result of feeding a single event to a projection session.</summary>
public sealed class FeedResult {
	/// <summary>Whether the event was processed or skipped.</summary>
	public required FeedStatus Status { get; init; }

	/// <summary>Partition key for the affected state. Null for skipped events and unpartitioned projections.</summary>
	public string? Partition { get; init; }

	/// <summary>Projection state for the affected partition after processing.</summary>
	public string? State { get; init; }

	/// <summary>Result for the affected partition. V1: transformed state (after <c>transformBy</c>/<c>filterBy</c>), or state if no transform; null if a filter excludes it. V2: post-handler state (transforms aren't invoked).</summary>
	public string? Result { get; init; }

	/// <summary>Shared state for biState projections, or null.</summary>
	public string? SharedState { get; init; }

	/// <summary>Events emitted during processing (emit/linkTo). Empty array when none.</summary>
	public EmittedEvent[] Emitted { get; init; } = [];

	/// <summary>Log messages from console.log calls. Empty array when none.</summary>
	public string[] Logs { get; init; } = [];

	/// <summary>Reason the event was skipped. Null when Status is Processed.</summary>
	public string? SkipReason { get; init; }

	/// <summary>Quirks encountered while processing this event (e.g. a biState string slot being JSON-quoted). Empty array when none. Distinct from compile-time <see cref="Gaffer.Sdk.ProjectionInfo"/> diagnostics: these are runtime, value-dependent, and have no source range.</summary>
	public Diagnostic[] Diagnostics { get; init; } = [];

	internal static FeedResult Skip(string reason) => new() { Status = FeedStatus.Skipped, SkipReason = reason };
}

/// <summary>Status of a feed operation.</summary>
public enum FeedStatus {
	/// <summary>Event was processed by the handler.</summary>
	Processed,
	/// <summary>Event was filtered out (V1 non-JSON, link, partitionBy null, etc).</summary>
	Skipped,
}

File: bindings/go/feed_result.go
package gafferruntime

import "encoding/json"

// FeedResult holds the result of feeding a single event to a projection.
type FeedResult struct {
	Status      string          `json:"status"`
	SkipReason  string          `json:"reason"`
	Partition   string          `json:"partition"`
	State       json.RawMessage `json:"state"`
	Result      json.RawMessage `json:"result"`
	SharedState json.RawMessage `json:"sharedState"`
	Emitted     []EmittedEvent  `json:"emitted"`
	Logs        []string        `json:"logs"`
	// Diagnostics holds quirks that fired while processing this event (e.g. a
	// biState string slot being JSON-quoted). Runtime, value-dependent, no
	// source range - distinct from compile-time ProjectionInfo.Diagnostics.
	Diagnostics []Diagnostic `json:"diagnostics"`
}

// EmittedEvent represents an event emitted by a projection via emit() or linkTo().
type EmittedEvent struct {
	StreamID  string            `json:"streamId"`
	EventType string            `json:"eventType"`
	Data      *string           `json:"data"`
	IsJson    bool              `json:"isJson"`
	IsLink    bool              `json:"isLink"`
	Metadata  map[string]string `json:"metadata"`
}
